lib/mosaic-geomath/src/main/java/org/eclipse/mosaic/lib/math/Matrix4d.java

```diff
@@ -15,92 +15,108 @@
-    }
-
-    public void dump() {
-        for (int m = 0; m < 4; m++) {
-            for (int n = 0; n < 4; n++) {
-                System.out.printf("%8.3f ", this.m[n * 4 + m]);
-            }
-            System.out.println();
-        }
+    /**
+     * Creates a new 4x4 matrix, with the values being copied from the given matrix.
+     */
+    public Matrix4d(Matrix4d copyFrom) {
+        set(copyFrom);
     }
 
-    public Matrix4d translate(Vector3d t) {
-        return translate(t.x, t.y, t.z);
-    }
-
-    public Matrix4d translate(double x, double y, double z) {
-        for (int i = 0; i < 4; i++) {
-            m[12 + i] += m[i] * x + m[4 + i] * y + m[8 + i] * z;
-        }
-        return this;
-    }
-
-    public Matrix4d rotate(double angleDeg, Vector3d axis) {
-        return rotate(angleDeg, axis.x, axis.y, axis.z);
-    }
-
-    public Matrix4d scale(double sx, double sy, double sz) {
-        for (int i = 0; i < 4; i++) {
-            m[i] *= sx;
-            m[4 + i] *= sy;
-            m[8 + i] *= sz;
-        }
-        return this;
-    }
-
-    public Matrix4d rotate(double angleDeg, double axX, double axY, double axZ) {
-        synchronized (tmpMatA) {
-            tmpMatA.setRotate(angleDeg, axX, axY, axZ);
-            set(multiply(tmpMatA, tmpMatB));
-        }
-        return this;
+    /**
+     * Creates a new identity matrix in 4x4 format.
+     */
+    public static Matrix4d identityMatrix() {
+        return new Matrix4d().setIdentity();
     }
 
+    /**
+     * Sets all values to match identity matrix.
+     */
     public Matrix4d setIdentity() {
         for (int i = 0; i < 16; i++) {
             m[i] = i % 5 == 0 ? 1 : 0;
         }
         return this;
     }
 
+    /**
+     * Sets all values to zero.
+     */
     public Matrix4d setZero() {
         for (int i = 0; i < 16; i++) {
             m[i] = 0;
         }
         return this;
     }
 
+    /**
+     * Copies the values from the given matrix to this matrix.
+     */
     public Matrix4d set(Matrix4d mat) {
-        for (int i = 0; i < 16; i++) {
-            m[i] = mat.m[i];
-        }
+        System.arraycopy(mat.m, 0, m, 0, 16);
+        return this;
+    }
+
+    /**
+     * Get a specific value from the matrix by row and column index
+     */
+    public double get(int row, int col) {
+        return m[row * 4 + col];
+    }
+
+    /**
+     * Sets a specific value in the matrix by row and column index
+     */
+    public Matrix4d set(int row, int col, double value) {
+        m[row * 4 + col] = value;
         return this;
     }
 
+    /**
+     * Adds a matrix to this matrix.
+     */
     public Matrix4d add(Matrix4d mat) {
+        return add(mat, this);
+    }
+
+    /**
+     * Adds a matrix to this matrix and writes the result into the result matrix.
+     */
+    public Matrix4d add(Matrix4d mat, Matrix4d result) {
         for (int i = 0; i < 16; i++) {
-            m[i] += mat.m[i];
+            result.m[i] += mat.m[i];
```

Comment:
same as for Matrix3d, either add `result.set(this)` or change to `result.m[i] = m[i] + mat.m[i];` (second one is better)